### PR TITLE
Don't use diagnostics memory pool in GracefulTurnsAbortiveIfRequestsDoNotFinish

### PR DESCRIPTION
**Pull request by JamesNK**
Addresses flaky test https://github.com/dotnet/aspnetcore/issues/39986

Abortive cancellation has undefined behavior and diagnostics memory pool throws errors in those (e.g. attempting to read memory after server and pool have shutdown). Use normal memory pool directly.

### CODE REVIEW

**Comment by JamesNK** [comment]
Review please 🙏

File: src/Servers/Kestrel/test/FunctionalTests/Http2/ShutdownTests.cs [hunk 1]
```diff
@@ -173,11 +167,9 @@ public async Task GracefulTurnsAbortiveIfRequestsDoNotFinish()
         var requestStarted = new TaskCompletionSource(TaskCreationOptions.RunContinuationsAsynchronously);
         var requestUnblocked = new TaskCompletionSource(TaskCreationOptions.RunContinuationsAsynchronously);
 
-        var memoryPoolFactory = new DiagnosticMemoryPoolFactory(allowLateReturn: true);
-
         var testContext = new TestServiceContext(LoggerFactory)
         {
-            MemoryPoolFactory = memoryPoolFactory.Create
+            MemoryPoolFactory = () => new PinnedBlockMemoryPool()
```
Comment (BrennanConroy) [inline src/Servers/Kestrel/test/FunctionalTests/Http2/ShutdownTests.cs:172]: Remove, this is set by default.

Comment (JamesNK) [reply src/Servers/Kestrel/test/FunctionalTests/Http2/ShutdownTests.cs:172]: In debug, a diagnostic memory pool is still created: https://github.com/dotnet/aspnetcore/blob/5cbfc712a4f726b4f16bc4fc35d945b57eb891de/src/Shared/Buffers.MemoryPool/MemoryPoolFactory.cs#L10-L14

It won't cause the CI build to fail (built with release) but explicitly not using a diagnostic memory pool prevents a flaky test in local test runs.

Comment (halter73) [reply src/Servers/Kestrel/test/FunctionalTests/Http2/ShutdownTests.cs:172]: Not exactly. `TestServiceContext.MemoryPoolFactory` defaults to `PinnedBlockMemoryPoolFactory.Create` which is the diagnostic pool in debug builds. That wouldn't affect the CI which runs release builds, but given that we expect abortive shutdown to sometimes lead to ODEs if abort takes longer than the non-configurable 1 second timeout, we shouldn't fail local tests on debug builds either.

**Review by halter73 — APPROVED** [review]

**Pull request merged**